### PR TITLE
CS-505: Marvin test to check VR internal DNS Service

### PR DESCRIPTION
**Pull request by abhinandanprateek**

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@abhinandanprateek can you check if `test_router_dns.py` is similar to your test, you may also move your test case in that test file? Thanks.

**Comment by abhinandanprateek** [comment]
@rhtyd the test_router_dns checks external dns service on virtual router. This test case checks the internal dns service from the user VMs.

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-55u3

File: test/integration/smoke/test_router_dnsservice.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            "Check list port forwarding rules"
+        )
+
+        result = None
+        try:
+            self.logger.debug("SSH into guest VM with IP: %s" % nat_rule1.ipaddress)
+            ssh = self.vm1.get_ssh_client(ipaddress=nat_rule1.ipaddress, port=self.services['natrule1']["publicport"], retries=8)
+            result = str(ssh.execute("ping drump"))
+            self.logger.debug("pinging drump: %s " % result)
+            result = str(ssh.execute("ping dilton"))
```
Comment (jburwell) [inline test/integration/smoke/test_router_dnsservice.py]: Please consider creating a common function to ping a host from another VM.  We seem to do it a lot, and a common function would be both more expressive and reduce duplication.

File: test/integration/smoke/test_router_dnsservice.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                                         templateid=cls.template.id,
+                                         accountid=cls.account.name,
+                                         domainid=cls.domain.id,
+                                         serviceofferingid=cls.service_offering.id,
+                                         networkids=[str(cls.network.id)])
+        cls.vm1.password = "password"
+        cls.logger.debug("Created VM named drump");
+                
+        cls.services["virtual_machine"]["displayname"] = 'dilton';
+        cls.services["virtual_machine"]["name"] = 'dilton';
```
Comment (jburwell) [inline test/integration/smoke/test_router_dnsservice.py]: Please consider extracting these hostnames to constants as the names are used throughout the test case.

File: test/integration/smoke/test_router_dnsservice.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+
+    def test_router_common(self):
+        """Performs common router tests and returns router public_ips"""
+
+        routers = list_routers(
+            self.apiclient,
+            account=self.account.name,
+            domainid=self.account.domainid
+        )
```
Comment (jburwell) [inline test/integration/smoke/test_router_dnsservice.py]: Please add an assertion that `routers` is not `None`.

File: test/integration/smoke/test_router_dnsservice.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        cls.network = Network.create(cls.api_client,
+                                      cls.services["network"],
+                                      accountid=cls.account.name,
+                                      domainid=cls.account.domainid,
+                                      networkofferingid=cls.network_offering.id,
+                                      zoneid=cls.zone.id)
+
+        cls.logger.debug("Creating guest VM for Account %s using offering %s" % (cls.account.name, cls.service_offering.id))
+        cls.services["virtual_machine"]["displayname"] = 'drump';
+        cls.services["virtual_machine"]["name"] = 'drump';
```
Comment (jburwell) [inline test/integration/smoke/test_router_dnsservice.py]: Please consider extracting these hostnames to constants as the names are used throughout the test case.

File: test/integration/smoke/test_router_dnsservice.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        nat_rule1 = NATRule.create(
+            self.apiclient,
+            self.vm1,
+            self.services["natrule1"],
+            public_ip.id
+        )
+        nat_rules = list_nat_rules(
+            self.apiclient,
+            id=nat_rule1.id
+        )
```
Comment (jburwell) [inline test/integration/smoke/test_router_dnsservice.py]: Please add an assertion that `nat_rules` is not `None`.

File: test/integration/smoke/test_router_dnsservice.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            self.logger.debug("pinging drump: %s " % result)
+            result = str(ssh.execute("ping dilton"))
+            self.logger.debug("pinging dilton: %s " % result)
+        except Exception as e:
+            self.fail("Failed to SSH into VM - %s due to exception: %s" % (nat_rule1.ipaddress, e))
+
+        if not result:
+            self.fail("Did not to receive any response from the guest VM, failing.")
+
+        self.assertTrue("dilton" in result and "#53" in result,
```
Comment (jburwell) [inline test/integration/smoke/test_router_dnsservice.py]: Where is the check of the result `ping drump` command?

**Comment by jburwell** [comment]
@murali-reddy is internal DNS service impacted by having multiple NICs?

File: test/integration/smoke/test_router_dnsservice.py (hunk 3 above)
Comment (abhinandanprateek) [inline test/integration/smoke/test_router_dnsservice.py]: This is followed by a set of assertions to ensure that that there is a router with valid state.

File: test/integration/smoke/test_router_dnsservice.py (hunk 5 above)
Comment (abhinandanprateek) [inline test/integration/smoke/test_router_dnsservice.py]: This is followed by a set of assertions to ensure that valid nat rules are there.

**Comment by abhinandanprateek** [comment]
@jburwell made the appropriate changes.
@murali-reddy impact of multiple NIC on DNS .... do you think we add a test for that too ?

**Comment by rhtyd** [comment]
@abhinandanprateek @murali-reddy please let me know any help needed?

**Comment by abhinandanprateek** [comment]
@rhtyd this can be merged if Murali feels multi nic test is also good to have we can add it later. cc @murali-reddy but lets not delay this as this test is complete in itself.

**Comment by murali-reddy** [comment]
@abhinandanprateek Multi-nic will be different scenario, we can skip that in this test case.

**Comment by rhtyd** [comment]
Test looks good to me, @abhinandanprateek should we target/include this in 4.9 as well?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
LGTM

Checks that guest VM can query VR DNS ... === TestName: test_router_dns_guestipquery | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 1 test in 277.198s

OK

**Pull request merged**